### PR TITLE
Download Speed Editorial

### PR DESCRIPTION
**Pull request by brebenelmihnea**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [ ] I have linked this PR to any issues that it closes.

### CODE REVIEW

**Comment by brebenelmihnea** [comment]
Why the entire flow module skips the Ford-Fulkerson algorithm ?

File: content/6_Advanced/Max_Flow.mdx [hunk 1]
```diff
@@ -10,6 +10,84 @@ frequency: 1
...
+			g[node][parent[node]] += curr_flow;
+			node = parent[node];
+		}
+		flow += curr_flow;
+		fill(parent.begin(), parent.end(), -1);
+	}
+	return flow;
+}
+
+long long capacity[NMAX][NMAX];
```
Comment (SansPapyrus683) [inline content/6_Advanced/Max_Flow.mdx:74]: can yo use a vector instead

File: content/6_Advanced/Max_Flow.mdx [hunk 2]
```diff
@@ -10,6 +10,84 @@ frequency: 1
...
+	int n, m;
+	cin >> n >> m;
+	for (int i = 0; i < m; i++) {
+		int x, y, c;
+		cin >> x >> y >> c;
+		x--, y--;
+		capacity[x][y] += c;
+	}
+
+	cout << max_flow(capacity, n, 0, n - 1) << '\n';
```
Comment (SansPapyrus683) [inline content/6_Advanced/Max_Flow.mdx:85]:
```suggestion
	cout << max_flow(capacity, n, 0, n - 1) << endl;
```

File: content/6_Advanced/Max_Flow.mdx (hunk 1 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Max_Flow.mdx]: Done

File: content/6_Advanced/Max_Flow.mdx (hunk 2 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Max_Flow.mdx]: Done

File: content/6_Advanced/Max_Flow.mdx [hunk 3]
```diff
@@ -10,6 +10,85 @@ frequency: 1
...
+			g[node][parent[node]] += curr_flow;
+			node = parent[node];
+		}
+		flow += curr_flow;
+		fill(parent.begin(), parent.end(), -1);
+	}
+	return flow;
+}
+
+vector<vector<long long>> capacity(NMAX, vector<long long>(NMAX, 0));
```
Comment (ryanchou-dev) [inline content/6_Advanced/Max_Flow.mdx:73]: move into loop and use n

File: content/6_Advanced/Max_Flow.mdx [hunk 4]
```diff
@@ -10,6 +10,85 @@ frequency: 1
...
+int main() {
+	int n, m;
+	cin >> n >> m;
+	for (int i = 0; i < m; i++) {
+		int x, y, c;
+		cin >> x >> y >> c;
+		x--, y--;
```
Comment (ryanchou-dev) [inline content/6_Advanced/Max_Flow.mdx:80]: sep lines

File: content/6_Advanced/Max_Flow.mdx [hunk 5]
```diff
@@ -10,6 +10,85 @@ frequency: 1
 
 <Problems problems="maxSam" />
 
+## Solution
+
+The code below uses the [Ford-Fulkerson](https://en.wikipedia.org/wiki/Ford%E2%80%93Fulkerson_algorithm) algorithm.
+
+The solution is a greedy approach. The download speed - the flow - can be improved as long as we can find a non-negative capacity path from the source - the server and the sink - Kotivalo's computer. After that we substract the minimum capacity on the found path from the edge capacities.
```
Comment (ryanchou-dev) [inline content/6_Advanced/Max_Flow.mdx:17]:
```suggestion

The solution is a greedy approach. The download speed (the flow) can be improved as long as we can find a non-negative capacity path from the source (the server) and the sink (Kotivalo's computer). After that, we subtract the minimum capacity on the found path from the edge capacities.
```

File: content/6_Advanced/Max_Flow.mdx [hunk 6]
```diff
@@ -10,6 +10,85 @@ frequency: 1
...
+
+using namespace std;
+
+const int NMAX = 500;
+
+long long max_flow(vector<vector<long long>> g, int n, int source, int sink) {
+	long long flow = 0;
+	vector<int> parent(n, -1);
+
+	// Find a wat from source to sink with non-negetive capacities
```
Comment (ryanchou-dev) [inline content/6_Advanced/Max_Flow.mdx:32]:
```suggestion
	// Find a way from source to sink with non-negative capacities
```

File: content/6_Advanced/Max_Flow.mdx [hunk 7]
```diff
@@ -10,6 +10,85 @@ frequency: 1
...
+
+		// The minimum capacity on the path from source to sink
+		long long curr_flow = LLONG_MAX;
+		while (node != source) {
+			curr_flow = min(curr_flow, g[parent[node]][node]);
+			node = parent[node];
+		}
+		node = sink;
+		while (node != source) {
+			// Substract the capacity from capacity edges
```
Comment (ryanchou-dev) [inline content/6_Advanced/Max_Flow.mdx:61]:
```suggestion
			// Subtract the capacity from capacity edges
```

File: content/6_Advanced/Max_Flow.mdx (hunk 4 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Max_Flow.mdx:80]: Ok

File: content/6_Advanced/Max_Flow.mdx (hunk 5 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Max_Flow.mdx:17]: Done

File: content/6_Advanced/Max_Flow.mdx (hunk 7 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Max_Flow.mdx:61]: Done

File: content/6_Advanced/Max_Flow.mdx (hunk 6 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Max_Flow.mdx:32]: Ok

File: content/6_Advanced/Max_Flow.mdx (hunk 3 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Max_Flow.mdx:73]: Done.

**Comment by brebenelmihnea** [comment]
@ryanchou-dev .

File: content/6_Advanced/Max_Flow.mdx [hunk 8]
```diff
@@ -10,6 +10,87 @@ frequency: 1
...
+		int x, y, c;
+		cin >> x >> y >> c;
+
+		x--, y--;
+		capacity[x][y] += c;
+	}
+
+	cout << max_flow(capacity, n, 0, n - 1) << endl;
+
+	return 0;
```
Comment (SansPapyrus683) [inline content/6_Advanced/Max_Flow.mdx:88]:
```suggestion
```

File: content/6_Advanced/Max_Flow.mdx (hunk 4 above)
Comment (SansPapyrus683) [reply content/6_Advanced/Max_Flow.mdx:82]: not fixed

File: content/6_Advanced/Max_Flow.mdx [hunk 9]
```diff
@@ -10,6 +10,87 @@ frequency: 1
...
+The solution is a greedy approach. The download speed (the flow) can be improved as long as we can find a non-negative capacity path from the source (the server) and the sink (Kotivalo's computer). After that we subtract the minimum capacity on the found path from the edge capacities.
+
+<LanguageSection>
+<CPPSection>
+```cpp
+#include <bits/stdc++.h>
+
+using namespace std;
+
+const int NMAX = 500;
```
Comment (ryanchou-dev) [inline content/6_Advanced/Max_Flow.mdx:26]: no longer needed

File: content/6_Advanced/Max_Flow.mdx [hunk 10]
```diff
@@ -10,6 +10,87 @@ frequency: 1
...
+
+long long max_flow(vector<vector<long long>> g, int n, int source, int sink) {
+	long long flow = 0;
+	vector<int> parent(n, -1);
+
+	// Find a way from source to sink with non-negetive capacities
```
Comment (ryanchou-dev) [inline content/6_Advanced/Max_Flow.mdx:32]: spelling

**Review by ryanchou-dev — COMMENTED** [review]
almost there!

File: content/6_Advanced/Max_Flow.mdx (hunk 8 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Max_Flow.mdx:88]: Done

File: content/6_Advanced/Max_Flow.mdx (hunk 9 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Max_Flow.mdx:26]: Yep

File: content/6_Advanced/Max_Flow.mdx (hunk 10 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Max_Flow.mdx]: Done

File: content/6_Advanced/Max_Flow.mdx (hunk 4 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Max_Flow.mdx:80]: Done

File: content/6_Advanced/Max_Flow.mdx [hunk 11]
```diff
@@ -10,6 +10,85 @@ frequency: 1
 
 <Problems problems="maxSam" />
 
+## Solution
+
+The code below uses the [Ford-Fulkerson](https://en.wikipedia.org/wiki/Ford%E2%80%93Fulkerson_algorithm) algorithm.
```
Comment (dongliu0426) [inline content/6_Advanced/Max_Flow.mdx:15]: your code uses edmonds-karp algorithm since you use a bfs instead of a dfs, right?

Comment (brebenelmihnea) [reply content/6_Advanced/Max_Flow.mdx]: Oh, you're right. I thought it was the other way around. Sorry !

**Review by dongliu0426 — APPROVED** [review]

**Review by SansPapyrus683 — APPROVED** [review]

**Review by ryanchou-dev — APPROVED** [review]

**Pull request merged**